From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phil Rodgers <philip.rodgers@glasgow.ac.uk>
Date: Wed, 21 Oct 2026 10:42:57 +0000
Subject: [PATCH 1/3] parser: keep pol integers as i64

`Token::Integer64` operands were stored as `isize` in
`PolToken::NonNegativeIntegerOrID`. On 32-bit targets this truncates pol
operands of 2^31 or more, so valid proofs are rejected (for example
tests/instances/correct/version3/general_watcher_bug_propagating_fully_at_level_zero).
The token now holds an `i64`; only its use as a constraint ID converts to
`isize`.
---
 veripb-parser/src/pbp_parser/parser.rs | 8 +++++---
 veripb-parser/src/pbp_parser/utils.rs  | 2 +-
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/veripb-parser/src/pbp_parser/parser.rs b/veripb-parser/src/pbp_parser/parser.rs
index 47c59b2..77de0f5 100644
--- a/veripb-parser/src/pbp_parser/parser.rs
+++ b/veripb-parser/src/pbp_parser/parser.rs
@@ -2753,7 +2753,7 @@ impl<'a, P: ProofProcessor, R: BufRead> Parser<'a, P, R> {
                     None
                 }
                 Token::Integer64(Sign::NoSign, id) => {
-                    Some(PolToken::NonNegativeIntegerOrID(id as isize))
+                    Some(PolToken::NonNegativeIntegerOrID(id))
                 }
                 token if self.is_constraint_id(&token)? => {
                     Some(PolToken::ConstraintId(self.constraint_id(token)?))
@@ -2841,10 +2841,12 @@ impl<'a, P: ProofProcessor, R: BufRead> Parser<'a, P, R> {
                 {
                     return Err(self.pol_parse_error(token, stack_size, operand).into());
                 }
-                (_, Some(PolToken::ConstraintId(id)))
-                | (_, Some(PolToken::NonNegativeIntegerOrID(id))) => {
+                (_, Some(PolToken::ConstraintId(id))) => {
                     instructions.push(Instruction::ConstraintId(id))
                 }
+                (_, Some(PolToken::NonNegativeIntegerOrID(id))) => {
+                    instructions.push(Instruction::ConstraintId(id as isize))
+                }
                 (_, Some(PolToken::Literal(literal))) => {
                     instructions.push(Instruction::LiteralAxiom(literal))
                 }
diff --git a/veripb-parser/src/pbp_parser/utils.rs b/veripb-parser/src/pbp_parser/utils.rs
index b046737..62a3f85 100644
--- a/veripb-parser/src/pbp_parser/utils.rs
+++ b/veripb-parser/src/pbp_parser/utils.rs
@@ -37,7 +37,7 @@ impl Scope {
 #[derive(Debug, PartialEq)]
 pub enum PolToken {
     ConstraintId(isize),
-    NonNegativeIntegerOrID(isize),
+    NonNegativeIntegerOrID(i64),
     PositiveInteger(BigInt),
     Literal(Lit),
 }
